HW3
f
i

NOTE : An assumption
In multiclass classification, we obtain a matrix of Beta's
Thus we get a matrix of overall 'parameters' despite our original vector of predictors as with each predictor, with each class a parameter is associated

THUS,
We won't attempt to select features in this case for each split. If parameter of a feature corresponding to one class is zero, the parameter corresponding to other class may not be. As such we could write a code to check if a feature has parameter zero in every class, if it does, then obviously it is to be dropped, however this does complify the problem


Thus L1-penalization will internally perform 'parameter-selection'
We could perform, based on the obtained parameters, 'feature elimination' and 'non-zero feature identification', however, this, here won't be performed

Thus we'll work with our base error_test_estimate vector and select the best l using the vector index where the error gets minimized. Then for that l we'll obtain the test set and obtain the actual test error, only and only for that l

In [2]:
import pandas as pd
import numpy as np
import math
from sklearn.linear_model import LogisticRegression
import statsmodels.api as sm 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import StratifiedKFold
from sklearn.linear_model import LogisticRegressionCV


df1 = pd.read_csv('bending1\\training\\dataset3.csv')
df2 = pd.read_csv('bending1\\training\\dataset4.csv')
df3 = pd.read_csv('bending1\\training\\dataset5.csv')
df4 = pd.read_csv('bending1\\training\\dataset6.csv')
df5 = pd.read_csv('bending1\\training\\dataset7.csv')




df6 = pd.read_csv('bending2\\training\\dataset3.csv')
df7 = pd.read_csv('bending2\\training\\dataset4.csv')
df8 = pd.read_csv('bending2\\training\\dataset5.csv')
df9 = pd.read_csv('bending2\\training\\dataset6.csv')
 



df10 = pd.read_csv('cycling\\training\\dataset4.csv')
df11 = pd.read_csv('cycling\\training\\dataset5.csv')
df12 = pd.read_csv('cycling\\training\\dataset6.csv')
df13 = pd.read_csv('cycling\\training\\dataset7.csv')
df14 = pd.read_csv('cycling\\training\\dataset8.csv')
df15 = pd.read_csv('cycling\\training\\dataset9.csv')
df16 = pd.read_csv('cycling\\training\\dataset10.csv')
df17 = pd.read_csv('cycling\\training\\dataset11.csv')
df18 = pd.read_csv('cycling\\training\\dataset12.csv')
df19 = pd.read_csv('cycling\\training\\dataset13.csv')
df20 = pd.read_csv('cycling\\training\\dataset14.csv')
df21 = pd.read_csv('cycling\\training\\dataset15.csv')






df22 = pd.read_csv('lying\\training\\dataset4.csv')
df23 = pd.read_csv('lying\\training\\dataset5.csv')
df24 = pd.read_csv('lying\\training\\dataset6.csv')
df25 = pd.read_csv('lying\\training\\dataset7.csv')
df26 = pd.read_csv('lying\\training\\dataset8.csv')
df27 = pd.read_csv('lying\\training\\dataset9.csv')
df28 = pd.read_csv('lying\\training\\dataset10.csv')
df29 = pd.read_csv('lying\\training\\dataset11.csv')
df30 = pd.read_csv('lying\\training\\dataset12.csv')
df31 = pd.read_csv('lying\\training\\dataset13.csv')
df32 = pd.read_csv('lying\\training\\dataset14.csv')
df33 = pd.read_csv('lying\\training\\dataset15.csv')






    
df34 = pd.read_csv('sitting\\training\\dataset4.csv')
df35 = pd.read_csv('sitting\\training\\dataset5.csv')
df36 = pd.read_csv('sitting\\training\\dataset6.csv')
df37 = pd.read_csv('sitting\\training\\dataset7.csv')
df38 = pd.read_csv('sitting\\training\\dataset8.csv')
df39 = pd.read_csv('sitting\\training\\dataset9.csv')
df40 = pd.read_csv('sitting\\training\\dataset10.csv')
df41 = pd.read_csv('sitting\\training\\dataset11.csv')
df42 = pd.read_csv('sitting\\training\\dataset12.csv')
df43 = pd.read_csv('sitting\\training\\dataset13.csv')
df44 = pd.read_csv('sitting\\training\\dataset14.csv')
df45 = pd.read_csv('sitting\\training\\dataset15.csv')







df46 = pd.read_csv('standing\\training\\dataset4.csv')
df47 = pd.read_csv('standing\\training\\dataset5.csv')
df48 = pd.read_csv('standing\\training\\dataset6.csv')
df49 = pd.read_csv('standing\\training\\dataset7.csv')
df50 = pd.read_csv('standing\\training\\dataset8.csv')
df51 = pd.read_csv('standing\\training\\dataset9.csv')
df52 = pd.read_csv('standing\\training\\dataset10.csv')
df53 = pd.read_csv('standing\\training\\dataset11.csv')
df54 = pd.read_csv('standing\\training\\dataset12.csv')
df55 = pd.read_csv('standing\\training\\dataset13.csv')
df56 = pd.read_csv('standing\\training\\dataset14.csv')
df57 = pd.read_csv('standing\\training\\dataset15.csv')










df58 = pd.read_csv('walking\\training\\dataset4.csv')
df59 = pd.read_csv('walking\\training\\dataset5.csv')
df60 = pd.read_csv('walking\\training\\dataset6.csv')
df61 = pd.read_csv('walking\\training\\dataset7.csv')
df62 = pd.read_csv('walking\\training\\dataset8.csv')
df63 = pd.read_csv('walking\\training\\dataset9.csv')
df64 = pd.read_csv('walking\\training\\dataset10.csv')
df65 = pd.read_csv('walking\\training\\dataset11.csv')
df66 = pd.read_csv('walking\\training\\dataset12.csv')
df67 = pd.read_csv('walking\\training\\dataset13.csv')
df68 = pd.read_csv('walking\\training\\dataset14.csv')
df69 = pd.read_csv('walking\\training\\dataset15.csv')






df1 = df1.drop(df1.columns[0],axis=1)
df2 = df2.drop(df2.columns[0],axis=1)
df3 = df3.drop(df3.columns[0],axis=1)
df4 = df4.drop(df4.columns[0],axis=1)
df5 = df5.drop(df5.columns[0],axis=1)
df6 = df6.drop(df6.columns[0],axis=1)
df7 = df7.drop(df7.columns[0],axis=1)
df8 = df8.drop(df8.columns[0],axis=1)
df9 = df9.drop(df9.columns[0],axis=1)
df10 = df10.drop(df10.columns[0],axis=1)
df11 = df11.drop(df11.columns[0],axis=1)
df12 = df12.drop(df12.columns[0],axis=1)
df13 = df13.drop(df13.columns[0],axis=1)
df14 = df14.drop(df14.columns[0],axis=1)
df15 = df15.drop(df15.columns[0],axis=1)
df16 = df16.drop(df16.columns[0],axis=1)
df17 = df17.drop(df17.columns[0],axis=1)
df18 = df18.drop(df18.columns[0],axis=1)
df19 = df19.drop(df19.columns[0],axis=1)
df20 = df20.drop(df20.columns[0],axis=1)
df21 = df21.drop(df21.columns[0],axis=1)
df22 = df22.drop(df22.columns[0],axis=1)
df23 = df23.drop(df23.columns[0],axis=1)
df24 = df24.drop(df24.columns[0],axis=1)
df25 = df25.drop(df25.columns[0],axis=1)
df26 = df26.drop(df26.columns[0],axis=1)
df27 = df27.drop(df27.columns[0],axis=1)
df28 = df28.drop(df28.columns[0],axis=1)
df29 = df29.drop(df29.columns[0],axis=1)
df30 = df30.drop(df30.columns[0],axis=1)
df31 = df31.drop(df31.columns[0],axis=1)
df32 = df32.drop(df32.columns[0],axis=1)
df33 = df33.drop(df33.columns[0],axis=1)
df34 = df34.drop(df34.columns[0],axis=1)
df35 = df35.drop(df35.columns[0],axis=1)
df36 = df36.drop(df36.columns[0],axis=1)
df37 = df37.drop(df37.columns[0],axis=1)
df38 = df38.drop(df38.columns[0],axis=1)
df39 = df39.drop(df39.columns[0],axis=1)
df40 = df40.drop(df40.columns[0],axis=1)
df41 = df41.drop(df41.columns[0],axis=1)
df42 = df42.drop(df42.columns[0],axis=1)
df43 = df43.drop(df43.columns[0],axis=1)
df44 = df44.drop(df44.columns[0],axis=1)
df45 = df45.drop(df45.columns[0],axis=1)
df46 = df46.drop(df46.columns[0],axis=1)
df47 = df47.drop(df47.columns[0],axis=1)
df48 = df48.drop(df48.columns[0],axis=1)
df49 = df49.drop(df49.columns[0],axis=1)
df50 = df50.drop(df50.columns[0],axis=1)
df51 = df51.drop(df51.columns[0],axis=1)
df52 = df52.drop(df52.columns[0],axis=1)
df53 = df53.drop(df53.columns[0],axis=1)
df54 = df54.drop(df54.columns[0],axis=1)
df55 = df55.drop(df55.columns[0],axis=1)
df56 = df56.drop(df56.columns[0],axis=1)
df57 = df57.drop(df57.columns[0],axis=1)
df58 = df58.drop(df58.columns[0],axis=1)
df59 = df59.drop(df59.columns[0],axis=1)
df60 = df60.drop(df60.columns[0],axis=1)
df61 = df61.drop(df61.columns[0],axis=1)
df62 = df62.drop(df62.columns[0],axis=1)
df63 = df63.drop(df63.columns[0],axis=1)
df64 = df64.drop(df64.columns[0],axis=1)
df65 = df65.drop(df65.columns[0],axis=1)
df66 = df66.drop(df66.columns[0],axis=1)
df67 = df67.drop(df67.columns[0],axis=1)
df68 = df68.drop(df68.columns[0],axis=1)
df69 = df69.drop(df69.columns[0],axis=1)






df_list = [df1,df2,df3,df4,df5,df6,df7,df8,df9,df10,df11,df12,df13,df14,df15,df16,df17,df18,df19,df20,df21,df22,df23,df24,df25,df26,df27,df28,df29,df30,df31,df32,df33,df34,df35,df36,df37,df38,df39,df40,df41,df42,df43,df44,df45,df46,df47,df48,df49,df50,df51,df52,df53,df54,df55,df56,df57,df58,df59,df60,df61,df62,df63,df64,df65,df66,df67,df68,df69]

out =     [0,  0,  0,  0,  0,  0,  0,  0,  0,  1,   1,    1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   2,   2,   2,   2,   2,  2,     2,    2,   2,   2,   2,  2,  3,   3,    3,    3,   3,  3,   3,   3,   3,   3,   3,   3,   4,    4,   4,  4,    4,   4,   4,   4,  4,   4,   4,   4,   5,   5,   5,   5,   5,   5,   5,   5,   5,   5,   5,   5]

# 69 datasets, each dataset will me split, features extracted, and instantiated in that same index in a corresponding manner
# thus out has 69 entries, corresponding to class in each dataset
# 0 : bending (1 & 2)
# 1 : cycling
# 2 : lying
# 3 : sitting
# 4 : standing
# 5 : walking
# ordered as can be seen again their dataframes, corresponding to df_list which will be instantiated in similar index orders
# thus this is a multiclass, 6 class problem, total 6 classes from 0 to 5

y = pd.DataFrame(out)



l = np.arange(1,21)
datalist = np.arange(0,69)
l_new = [1,1,2,2,3,3,4,4,5,5,6,6,7,7,8,8,9,9,10,10,11,11,12,12,13,13,14,14,15,15,16,16,17,17,18,18,19,19,20,20]
errors_test_estimated = pd.DataFrame(index=l, columns = ['Estimated Test Error'])


for i in l:
    
    splits = np.arange(0,i)
    segment = math.floor(480/i)
    df_list_new = []
    j = 0
    for j in datalist:
        list_df_specific = []
        df_parts = 0
        for df_parts in splits:
             list_df_specific.append(df_list[j].iloc[((df_parts)*(segment)) : ((df_parts + 1)*(segment)),:].reset_index(drop=True))
        df_list_new.append(pd.concat(list_df_specific, axis=1))
    
    header_column = np.arange(0, 18*i) 
    features = pd.DataFrame(index=datalist,columns=header_column) 
   

    

    t=0
    for t in datalist:
        features.iloc[t, 0 : (6*i)]       =    (((pd.DataFrame(df_list_new[t].mean()).transpose()).to_numpy()).flatten())[:]
        features.iloc[t, (6*i) : (12*i)]  =    (((pd.DataFrame(df_list_new[t].std()).transpose()).to_numpy()).flatten())[:]
        features.iloc[t, (12*i) : (18*i)] =    (((pd.DataFrame(df_list_new[t].max()).transpose()).to_numpy()).flatten())[:]
    
    
    total = features.copy()
    total['Output'] = out
    
    # note this out will stay same, irrespective of split, as each spit in each dataset in markedly instantiated against
    # its corresponding out
    # thus no matter the split index row 0 will have instantiated dataset df1 of bending 1 which against it will contain its
    # class 0, also in corresponding index row 0
    # and so on and so forth for all the corresponding classes in general
    
    # total has split features with output class, this is only the training data
    
    # total has the relevant split data that has to be cross validated upon
    
    skf = StratifiedKFold(n_splits=5,shuffle=True)  # performing 5-fold CV 
    
    X_train = total.drop(columns=['Output'])
    y_train = total['Output']                   
    
    
    # the errors will be saved in the error_test_estimate vector
    # here external cv is stratified, as it wasn't used before we'll use it here
    # additionally internal cv is automatically stratified by sklearn
    
    cv_error = np.zeros(5) #to store cv errors
    
    
    count_ext = 0
    for train_index, test_index in skf.split(X_train, y_train):
        X_train_cv,X_test_cv = X_train.iloc[train_index,:],X_train.iloc[test_index,:]
        y_train_cv,y_test_cv = y_train.iloc[train_index],y_train.iloc[test_index]
        count_ext = count_ext + 1
        log_reg = LogisticRegressionCV(cv=5,solver='liblinear',penalty='l1').fit(X_train_cv, y_train_cv)
        
        # 5-fold CV picks best lambda/alpha/C 
        
        # note liblinear and sklearn penalizes beta0 also, by default, so it will be considered a 'parameter'

        y_test_cv_predicted = log_reg.predict(X_test_cv)
        
        mis = 0
        h=0
        for h in np.arange(0,y_test_cv.shape[0]):
            if y_test_cv_predicted[h] != y_test_cv.iloc[h]:
                mis = mis + 1
        
        error = (mis/(y_test_cv.shape[0]))*100
        
        cv_error[count_ext-1]=error
        
    
    
    
   
   
    error_l = np.mean(cv_error)
    errors_test_estimated.iloc[i-1,0]=error_l
    
    
    
print('The test error estimates : \n',errors_test_estimated)


C:\Users\DHRUV\anaconda3\lib\site-packages\sklearn\svm\_base.py:977: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\DHRUV\anaconda3\lib\site-packages\sklearn\svm\_base.py:977: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\DHRUV\anaconda3\lib\site-packages\sklearn\svm\_base.py:977: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\DHRUV\anaconda3\lib\site-packages\sklearn\svm\_base.py:977: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\DHRUV\anaconda3\lib\site-packages\sklearn\svm\_base.py:977: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iteration

The test error estimates : 
    Estimated Test Error
1               15.9341
2               13.0769
3               11.5385
4               24.7253
5               17.4725
6                21.978
7                22.967
8               24.6154
9               24.5055
10              30.3297
11              23.0769
12              23.4066
13              30.5495
14              27.8022
15              27.4725
16              23.2967
17               31.978
18               29.011
19              27.4725
20               26.044


In [5]:
index = np.argmin(errors_test_estimated)
l_selected = index + 1
print('The l split selected that minimizes the test error estimate for this multiclass classification is : \n',l_selected)

The l split selected that minimizes the test error estimate for this multiclass classification is : 
 3


Comments :
As can be seen, the total errors are much higher than for binary classification. This could be expected since the overall instances for each class now has reduced. 
NOTE : Stratified cross validation was used here as we needed to preserve the existence of each class label in each split, otherwise error estimated would be highly overestimated as otherwise certain classes would appear in train and would be there in test

Thus stratified Kfold utilization is of utmost importance here in general
l=3 has minimum test error estimate

FOR actual TEST ERROR :
Assumption : Since each run may yield a different l-value as kfold is completely randomized and uses shuffled data, the l value in each run may be different. In general, it should vary mostly between 1-3 (1,2,3 splits) as that consistently across all problems minimized the test error (estimate obviously, via Kfold)

Here we have two routes :
1. Take the selected l and split according to that l so that variation in l across run time (minor) will ensure that test data is split according to that selected l
2. Sectionalize and assume that l=3 is the answer, thus if at a different run time we get l=1, we'll still split data according to current run time result, that is in 3 parts

While coding approach 1. is more general, it is again mildly time consuming and repetitive in terms of re-splitting according to the selected l. Obviously 1. isn't that difficult or doesn't need additonal coding or more code blocks, just a code block that splits everytime according to the obtained l
In general, we'll take approach 2 into consideration here and go forth with that, as opposed to approach 1.

THUS
NOTE : Runtime change may give different l, we'll assume l=3 is sufficiently good estimate for l over large number of trials

Thus we will split test into 3 parts, and obtain the test error on it by training it on train data, that we will also split once again

In [6]:
i=3       # train split into 3
    
splits = np.arange(0,i)
segment = math.floor(480/i)
df_list_new = []
j = 0
for j in datalist:
    list_df_specific = []
    df_parts = 0
    for df_parts in splits:
         list_df_specific.append(df_list[j].iloc[((df_parts)*(segment)) : ((df_parts + 1)*(segment)),:].reset_index(drop=True))
    df_list_new.append(pd.concat(list_df_specific, axis=1))

header_column = np.arange(0, 18*i) 
features = pd.DataFrame(index=datalist,columns=header_column) 




t=0
for t in datalist:
    features.iloc[t, 0 : (6*i)]       =    (((pd.DataFrame(df_list_new[t].mean()).transpose()).to_numpy()).flatten())[:]
    features.iloc[t, (6*i) : (12*i)]  =    (((pd.DataFrame(df_list_new[t].std()).transpose()).to_numpy()).flatten())[:]
    features.iloc[t, (12*i) : (18*i)] =    (((pd.DataFrame(df_list_new[t].max()).transpose()).to_numpy()).flatten())[:]
    
total = features.copy()
total['Output'] = out

print(total)


          0         1        2         3        4         5        6  \
0   43.9451  0.350125  23.3911     0.455  36.3489  0.530687  44.3242   
1   43.6212   0.59475  23.1458   0.71375  36.5305  0.654437  42.1004   
2   40.6738  0.698312  18.8588   1.25144  29.0874  0.439375  42.8464   
3   44.7921  0.177937  18.9189   1.11831  24.0943    0.8115  42.8199   
4   43.0673   0.40875  15.4131  0.825563  23.4773  0.522188  44.6137   
..      ...       ...      ...       ...      ...       ...      ...   
64  33.3256   4.37131   15.106   3.25338  15.8803   3.23538  33.6204   
65  34.1833   4.85762  15.0886   3.26744  15.4443   3.42537  34.2593   
66  34.0153   4.62231  15.5089   3.12131  16.1948   3.29656  34.5384   
67  34.8948    4.4205  15.5336   3.12975  15.6431   3.44369  34.8862   
68  35.0022   4.28537  15.1794   3.20906  15.6646   3.29419  34.2569   

           7        8         9  ...    45     46    47     48     49     50  \
0   0.408062  22.3036     0.623  ...  4.44  38.33  1.79

In [8]:
X_train = total.drop(columns = ['Output'])
y_train = total['Output']
print(X_train)
print(y_train)

         0         1        2         3        4         5        6   \
0   43.9451  0.350125  23.3911     0.455  36.3489  0.530687  44.3242   
1   43.6212   0.59475  23.1458   0.71375  36.5305  0.654437  42.1004   
2   40.6738  0.698312  18.8588   1.25144  29.0874  0.439375  42.8464   
3   44.7921  0.177937  18.9189   1.11831  24.0943    0.8115  42.8199   
4   43.0673   0.40875  15.4131  0.825563  23.4773  0.522188  44.6137   
..      ...       ...      ...       ...      ...       ...      ...   
64  33.3256   4.37131   15.106   3.25338  15.8803   3.23538  33.6204   
65  34.1833   4.85762  15.0886   3.26744  15.4443   3.42537  34.2593   
66  34.0153   4.62231  15.5089   3.12131  16.1948   3.29656  34.5384   
67  34.8948    4.4205  15.5336   3.12975  15.6431   3.44369  34.8862   
68  35.0022   4.28537  15.1794   3.20906  15.6646   3.29419  34.2569   

          7        8         9   ...     44    45     46    47     48     49  \
0   0.408062  22.3036     0.623  ...  29.25  4.44  38.3

In [14]:
df_n_1 = pd.read_csv('bending1\\test\\dataset1.csv')
df_n_2 = pd.read_csv('bending1\\test\\dataset2.csv')

df_n_3 = pd.read_csv('bending2\\test\\dataset1.csv')
df_n_4= pd.read_csv('bending2\\test\\dataset2.csv')

df_n_5 = pd.read_csv('cycling\\test\\dataset1.csv')
df_n_6 = pd.read_csv('cycling\\test\\dataset2.csv')
df_n_7 = pd.read_csv('cycling\\test\\dataset3.csv')

df_n_8 = pd.read_csv('lying\\test\\dataset1.csv')
df_n_9 = pd.read_csv('lying\\test\\dataset2.csv')
df_n_10 = pd.read_csv('lying\\test\\dataset3.csv')

df_n_11 = pd.read_csv('sitting\\test\\dataset1.csv')
df_n_12 = pd.read_csv('sitting\\test\\dataset2.csv')
df_n_13 = pd.read_csv('sitting\\test\\dataset3.csv')

df_n_14 = pd.read_csv('standing\\test\\dataset1.csv')
df_n_15 = pd.read_csv('standing\\test\\dataset2.csv')
df_n_16 = pd.read_csv('standing\\test\\dataset3.csv')

df_n_17 = pd.read_csv('walking\\test\\dataset1.csv')
df_n_18 = pd.read_csv('walking\\test\\dataset2.csv')
df_n_19 = pd.read_csv('walking\\test\\dataset3.csv')

df_n_1 = df_n_1.drop(df_n_1.columns[0],axis=1)
df_n_2 = df_n_2.drop(df_n_2.columns[0],axis=1)
df_n_3 = df_n_3.drop(df_n_3.columns[0],axis=1)
df_n_4 = df_n_4.drop(df_n_4.columns[0],axis=1)
df_n_5 = df_n_5.drop(df_n_5.columns[0],axis=1)
df_n_6 = df_n_6.drop(df_n_6.columns[0],axis=1)
df_n_7 = df_n_7.drop(df_n_7.columns[0],axis=1)
df_n_8 = df_n_8.drop(df_n_8.columns[0],axis=1)
df_n_9 = df_n_9.drop(df_n_9.columns[0],axis=1)
df_n_10 = df_n_10.drop(df_n_10.columns[0],axis=1)
df_n_11 = df_n_11.drop(df_n_11.columns[0],axis=1)
df_n_12 = df_n_12.drop(df_n_12.columns[0],axis=1)
df_n_13 = df_n_13.drop(df_n_13.columns[0],axis=1)
df_n_14 = df_n_14.drop(df_n_14.columns[0],axis=1)
df_n_15 = df_n_15.drop(df_n_15.columns[0],axis=1)
df_n_16 = df_n_16.drop(df_n_16.columns[0],axis=1)
df_n_17 = df_n_17.drop(df_n_17.columns[0],axis=1)
df_n_18 = df_n_18.drop(df_n_18.columns[0],axis=1)
df_n_19 = df_n_19.drop(df_n_19.columns[0],axis=1)

df_list_test = [df_n_1,df_n_2,df_n_3,df_n_4,df_n_5,df_n_6,df_n_7,df_n_8,df_n_9,df_n_10,df_n_11,df_n_12,df_n_13,df_n_14,df_n_15,df_n_16,df_n_17,df_n_18,df_n_19]

out_test =     [0,      0,      0,     0,     1,     1,     1,     2,     2,     2,     3,        3,      3,     4,      4,     4,       5,      5,       5]

y_test = pd.DataFrame(out_test)

datalist_test = np.arange(0,19)



i=3       # test split into 3
    
splits = np.arange(0,i)
segment = math.floor(480/i)
df_list_new = []
j = 0
for j in datalist_test:
    list_df_specific = []
    df_parts = 0
    for df_parts in splits:
         list_df_specific.append(df_list_test[j].iloc[((df_parts)*(segment)) : ((df_parts + 1)*(segment)),:].reset_index(drop=True))
    df_list_new.append(pd.concat(list_df_specific, axis=1))

header_column = np.arange(0, 18*i) 
features_test = pd.DataFrame(index=datalist_test,columns=header_column) 




t=0
for t in datalist_test:
    features_test.iloc[t, 0 : (6*i)]       =    (((pd.DataFrame(df_list_new[t].mean()).transpose()).to_numpy()).flatten())[:]
    features_test.iloc[t, (6*i) : (12*i)]  =    (((pd.DataFrame(df_list_new[t].std()).transpose()).to_numpy()).flatten())[:]
    features_test.iloc[t, (12*i) : (18*i)] =    (((pd.DataFrame(df_list_new[t].max()).transpose()).to_numpy()).flatten())[:]
    
total_test = features_test.copy()
total_test['Output'] = out_test

print(total_test)



          0         1        2         3        4         5        6  \
0   41.3626  0.256125  18.3271  0.581188  34.0511    0.6515  39.6206   
1   42.9217  0.468438  19.9269  0.834688  33.4829  0.505687  41.7014   
2   22.6099  0.871188   19.152  0.603438  25.9284  0.536312  25.7496   
3   28.7929  0.211625  19.5479  0.613125  17.3392  0.863625  28.6321   
4   36.7948   2.45981  17.5693   3.06381  19.1185   2.95456  37.3584   
5   37.5218    2.1565  16.0797   3.28169  18.7794    2.9165  37.5402   
6   38.0289   2.00513  16.3569   2.69894  19.7927   2.49438  37.0514   
7   27.7881  0.234312  6.04856     0.846    9.313  0.607875   26.624   
8   36.5416  0.291875  6.68925  0.904063  7.79194  0.960563  48.0072   
9        48         0  5.61125  0.397187  3.78381  0.696625  48.0016   
10  42.3121   0.29425  16.3537  0.729625  9.86094   1.26156  43.4542   
11  45.9493  0.213687  13.6939  0.787188  18.2038  0.801688  46.0373   
12  45.5889  0.349562  14.6733   1.13175  18.8676   1.08931  45.

In [16]:
X_test = total_test.drop(columns=['Output'])
y_test = total_test['Output']
print(X_test)
print(y_test)

         0         1        2         3        4         5        6   \
0   41.3626  0.256125  18.3271  0.581188  34.0511    0.6515  39.6206   
1   42.9217  0.468438  19.9269  0.834688  33.4829  0.505687  41.7014   
2   22.6099  0.871188   19.152  0.603438  25.9284  0.536312  25.7496   
3   28.7929  0.211625  19.5479  0.613125  17.3392  0.863625  28.6321   
4   36.7948   2.45981  17.5693   3.06381  19.1185   2.95456  37.3584   
5   37.5218    2.1565  16.0797   3.28169  18.7794    2.9165  37.5402   
6   38.0289   2.00513  16.3569   2.69894  19.7927   2.49438  37.0514   
7   27.7881  0.234312  6.04856     0.846    9.313  0.607875   26.624   
8   36.5416  0.291875  6.68925  0.904063  7.79194  0.960563  48.0072   
9        48         0  5.61125  0.397187  3.78381  0.696625  48.0016   
10  42.3121   0.29425  16.3537  0.729625  9.86094   1.26156  43.4542   
11  45.9493  0.213687  13.6939  0.787188  18.2038  0.801688  46.0373   
12  45.5889  0.349562  14.6733   1.13175  18.8676   1.08931  45.

In [17]:
cls = LogisticRegressionCV(cv=5,solver='liblinear',penalty='l1').fit(X_train, y_train)
y_test_predicted = cls.predict(X_test)    
print(y_test)
print(y_test_predicted)

0     0
1     0
2     0
3     0
4     1
5     1
6     1
7     2
8     2
9     2
10    3
11    3
12    3
13    4
14    4
15    4
16    5
17    5
18    5
Name: Output, dtype: int64
[0 0 0 4 1 1 5 2 4 2 4 3 3 2 2 4 5 5 5]


In [19]:
wrng = 0
for q in np.arange(0,y_test.shape[0]):
    if y_test_predicted[q] != y_test.iloc[q]:
        wrng = wrng + 1

wrngper = (wrng/(y_test.shape[0]))*100
print('The actual test error is : \n',wrngper,'%')

The actual test error is : 
 31.57894736842105 %


Thus, despite being trained on the entire training data in the current case, the actual test error obtained was 31 %, much higher than estimated test error 11 %

For my Run :
Best Split, l = 3
Test Error = 31 %
Estimated Test Error = 11 %

Based on runs l could change, and thus estimated error for test corresponding to an l could change. This is probably why in the current case, the test error is so different from its estimate

One of the reasons why this happens here more so is because despite having 6 classes, the dataset contains only 69 instances to train the dataset and the test dataset is comparable in size to the train

Thus we have such fluidity in test errors, their estimates, and obtained optimal l, but more often than not l will fall in between 1,2,3 splits

Since I don't have R, I can't use pROC
I've done ROC manually for the previous portions, but this would clearly be time consuming here
Thus, I've chosen to not perform the :
Confusion Matrix, ROC, AUC for this multi-class classification
As such we could use either the macro or the micro techniques of estimating them as discussed in the lectures, and use manual methods rather than pROC to do these steps. It shouldn't be difficult or that different, coding wise, just time consuming